File: sol/2900/2900.cpp
struct Solution {
    vector<string> getLongestSubsequence(vector<string>& words, vector<int>& groups) {
        int n = words.size();
        vector<int> prevIdx(n, -1), seqLen(n + 1, 1);
        seqLen[0] = 0;
        int lastOccur[2] = {-1, -1};
        for (int i = 0; i < n; ++i){
            if (lastOccur[1 - groups[i]] != -1){
                prevIdx[i] = lastOccur[1 - groups[i]];
                seqLen[i + 1] = seqLen[lastOccur[1 - groups[i]] + 1] + 1;
            }
            lastOccur[groups[i]] = i;
        }
        int gp = (seqLen[lastOccur[1] + 1] >= seqLen[lastOccur[0] + 1]) ? 1 : 0;
        int ptr = lastOccur[gp];
        vector<string> res(seqLen[ptr + 1]);
        for (int i = seqLen[ptr + 1] - 1; i >= 0; --i){
            res[i] = words[ptr];
            ptr = prevIdx[ptr];
        }
        return res;
    }
};

File: sol/2900/2900.py
class Solution:
    def getLongestSubsequence(self, words: List[str], groups: List[int]) -> List[str]:
        n = len(words)
        prevIdx = [-1] * n
        lastOccur = [-1, -1]
        seqLen = [1] * (n + 1)
        seqLen[-1] = 0
        for i, c in enumerate(groups):
            if lastOccur[1 - c] != -1:
                prevIdx[i] = lastOccur[1 - c]
                seqLen[i] = seqLen[lastOccur[1 - c]] + 1
            lastOccur[c] = i
        gp = 1 if seqLen[lastOccur[1]] >= seqLen[lastOccur[0]] else 0
        ptr = lastOccur[gp]
        buff = []
        while ptr != -1:
            buff.append(words[ptr])
            ptr = prevIdx[ptr]
        return buff[::-1]
